File: python/SelfGroups.py
# coding: utf-8
# SelfGroups.py
# 同じSelfFillTypeのRecordの集まり

from __future__ import annotations
from typing import Optional, Iterator, List, Tuple
from itertools import *

from VCFImpSelfRecord import SelfFillType, VCFImpSelfRecord
from VCFSelfFillableRecord import VCFSelfFillableRecord
from SelfRecordSet import SelfRecordSet

Group = Tuple[SelfFillType, List[VCFSelfFillableRecord]]


#################### SelfGroups ####################

class SelfGroups:
	def __init__(self, groups: list[Group]):
		self.groups: list[Group] = groups
	
	def __len__(self) -> int:
		return len(self.groups)
	
	def find_prev_record(self, i: int, g: SelfFillType
								) -> Optional[VCFSelfFillableRecord]:
		key, records = self.groups[i]
		chr = records[0].v[0]
		for j in range(i-1, -1, -1):
			key, records = self.groups[j]
			if records[0].v[0] != chr:
				return None
			if key == g:
				return records[-1]
		else:
			return None
	
	def find_next_record(self, i: int, g: SelfFillType
									) -> Optional[VCFSelfFillableRecord]:
		key, records = self.groups[i]
		chr = records[0].v[0]
		for j in range(i+1, len(self)):
			key, records = self.groups[j]
			if records[0].v[0] != chr:
				return None
			if key == g:
				return records[0]
		else:
			return None
	
	def generate_record_sets(self) -> Iterator[SelfRecordSet]:
		for i, (key, records) in enumerate(self.groups):
			if key == SelfFillType.P01 or key == SelfFillType.FILLED:
				continue
			prev_record = self.find_prev_record(i, SelfFillType.P01)
			next_record = self.find_next_record(i, SelfFillType.P01)
			for record in records:
				yield SelfRecordSet(record, prev_record, next_record)
	
	@staticmethod
	def create(records: list[VCFSelfFillableRecord]) -> SelfGroups:
		groups: list[Group] = []
		for g, v in groupby(records, key=lambda r: r.type):
			groups.append((g, list(v)))
		return SelfGroups(groups)

File: src/SelfGroups.cpp
// SelfGroups.cpp
#include "../include/SelfGroups.h"
#include "../include/VCFSelfFillableRecord.h"

using namespace std;


//////////////////// SelfGroups ////////////////////

VCFSelfFillableRecord *SelfGroups::find_prev_record(size_t i,
												SelfFillType g) const {
	if(i == 0)
		return NULL;
	
	for(size_t j = i-1; j < i; --j) {	// underflow
		const SelfFillType	key = get_group(j).first;
		const auto&	records = get_group(j).second;
		if(key == g)
			return records.back();
	}
	return NULL;
}

VCFSelfFillableRecord *SelfGroups::find_next_record(size_t i,
												SelfFillType g) const {
	for(size_t j = i+1; j < size(); ++j) {
		const SelfFillType	key = get_group(j).first;
		const auto&	records = get_group(j).second;
		if(key == g)
			return records.front();
	}
	return NULL;
}

vector<SelfRecordSet *> SelfGroups::create_record_sets() const {
	vector<SelfRecordSet *>	record_sets;
	for(size_t i = 0; i < size(); ++i) {
		const SelfFillType	key = get_group(i).first;
		const auto&	records = get_group(i).second;
		if(key == SelfFillType::P01 || key == SelfFillType::FILLED)
			continue;
		auto	*prev_record = find_prev_record(i, SelfFillType::P01);
		auto	*next_record = find_next_record(i, SelfFillType::P01);
		for(auto p = records.begin(); p != records.end(); ++p) {
			record_sets.push_back(new SelfRecordSet(*p, prev_record,
														next_record));
		}
	}
	return record_sets;
}

SelfGroups *SelfGroups::create(const vector<VCFSelfFillableRecord *>& records) {
	vector<Group>	groups;
	auto	current_type = records.front()->get_type();
	vector<VCFSelfFillableRecord *>	group(1U, records.front());
	for(auto p = records.begin() + 1; p != records.end(); ++p) {
		auto	*record = *p;
		if(record->get_type() != current_type) {
			groups.push_back(Group(current_type, group));
			group.clear();
			current_type = record->get_type();
		}
		group.push_back(record);
	}
	groups.push_back(Group(current_type, group));
	return new SelfGroups(groups);
}
